### PR TITLE
Integrate spotless

### PR DESCRIPTION
**Pull request by blacelle**
### SUMMARY
- Integrate Spotless (https://github.com/diffplug/spotless/issues/2321)
- (Takes over https://github.com/RoaringBitmap/RoaringBitmap/pull/758)

## Task list

- Uncomment ratchetFrom before merging in master
- Ensure ./gradlew test to apply `spotlessCheck` (or change the documentation)
- Add a CI step doing `spotlessCheck`

### Automated Checks

- [ ] I have run `./gradlew test` and made sure that my PR does not break any unit test.

### CODE REVIEW

File: bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java
Comment (blacelle) [inline bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java:9]: @lemire Are you fine with such Javadoc reformatting?

File: bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java
Comment (blacelle) [inline bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java:61]: @lemire Do you feel something is wrong with some max-width parameter? (i.e. is this reformatting expected, or should I dig for maxWidth parameter?)

**Comment by blacelle** [comment]
@lemire This demonstrates the whole codebase post `./gradlew spotlessApply`. 

We may prefer not such a full reformat, apply `ratcherFrom` from the start, and let PR reformatting the code through time. Your call. (I would advise a full-reformat, else people may complain their PR impact more than what they expected to contribute).

**Comment by lemire** [comment]
Let us ping recent contributors:

@kf-vikas  @fanson @anandheritage @iszhangpch  @zoltanmeze 


And also core contributors like @richardstartin 

If anyone objects to a full code reformat, please speak up as this can be quite damaging if you have forked the project and are in the process of changing it...

There will be no going back once this PR is applied.

File: bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java
Comment (lemire) [reply bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java:9]: That's not entirely great, is it?

File: bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java
Comment (lemire) [inline bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java:39]: @blacelle That's kind of messed up... the comment is no longer understandable because of the reformat.

Is there a way to tell it not to reformat the comments beyond, maybe, breaking long lines?

File: bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java
Comment (lemire) [reply bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java:61]: That particular change does not offend me, but

Comment (lemire) [reply bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java:61]: I think we  should avoid reformatting the comments too much.

File: bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java
Comment (nedtwigg) [reply bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java:39]: you can do `googleJavaFormat().formatJavadoc(false)`

the underlying issue is that javadoc is supposed to be html-ish, I think the rendered HTML Javadoc here will also look weird, and it would ideally be `<ol><li>` etc. If you put in the right HTML tags, then I think gjf will format it nicely too. In this case I think the formatting is sort of "catching a bug", but perfectly reasonable to ignore it too.

**Comment by lemire** [comment]
@blacelle Can you do the same PR but with `googleJavaFormat().formatJavadoc(false)`? I take @nedtwigg's point that JavaDoc should be HTML-like, but that's not the case right now and we don't want messed up comments.

**Comment by blacelle** [comment]
```
git reset --hard 14817b115021909e696f05f8512837f795164dee
git checkout 9d9dbe54a14dc53b07eed62a2e1aa885f76b751e build.gradle.kts
git commit XXX
./gradlew spotlessApply
git commit YYY
```

@lemire The PR is updated with javadoc formatting disabled. There is still many impacted files and LOC.

**Comment by lemire** [comment]
@blacelle 


I had a quick look and it seems fine to me.

I tried pinging a lot of folks, let us give them time to react. If nobody objects, and you are happy with the PR, let us merge it.

**Comment by lemire** [comment]
Great work. Let us give time for people to object, if they want.

**Comment by zoltanmeze** [comment]
I think it's reasonable to do a full code reformat. Just make sure the formatting rules are enforced on new PRs.

Personally, I prefer checkstyle analysis with somewhat stricter rules over spotless.
But here it makes sense to use spotless auto format feature, as it doesn't require any additional setup.

By looking at the changes, there are couple of things I'm not a huge fan of:
* wildcard imports
* not enforcing curly braces - could easily lead to a bug
* custom import order - prefer internal packages first, makes it easier to understand the code structure/ hierarchy

But that's just my personal opinion, and I don't think there is an out-of-box solution for the first two.

For custom imports, I would suggest including (if anyone else agrees):
```kotlin
importOrder("\\#", "org.roaringbitmap", "", "java", "javax")
```


> If anyone objects to a full code reformat, please speak up as this can be quite damaging if you have forked the project and are in the process of changing it...

I don't think it's that damaging, just choose local changes on conflicts and run `./gradlew spotlessApply` after.

File: build.gradle.kts
Comment (anandheritage) [inline build.gradle.kts:8]: Can we have a stable version instead of Beta version ? 
Stable version to avoid potential issues down the line.

Comment (lemire) [reply build.gradle.kts:8]: We could roll back to [6.24.0](https://github.com/diffplug/spotless/releases/tag/gradle%2F6.24.0)

**Comment by blacelle** [comment]
> not enforcing curly braces - could easily lead to a bug

Might be doable with Eclipse formatter, but I suppose we'd rather sticking with GoogleStandard, than maintaining an Eclipse stylesheet.

> wildcard imports

https://github.com/diffplug/spotless/issues/649
I added some conf, replicating what's done in OpenSearch (over Spotless) not to clean them, but to prevent them.

**Review by zoltanmeze — APPROVED** [review]
If no one else objects, I'm happy with the current format. Great work!

**Comment by lemire** [comment]
The trend is that the community approves this PR. Let us give it a few more days...

**Comment by kf-vikas** [comment]
> Let us ping recent contributors:
> 
> @kf-vikas @fanson @anandheritage @iszhangpch @zoltanmeze
> 
> And also core contributors like @richardstartin
> 
> If anyone objects to a full code reformat, please speak up as this can be quite damaging if you have forked the project and are in the process of changing it...
> 
> There will be no going back once this PR is applied.

Sorry, a bit late to catch up. 

I was hoping that the PR will contain a summary describing motivation/ existing problems for making this change.

Since this change is too big, so it is beyond human review. The only way to review is to ensure that this commit hash matches from where it is picked up and it worked originally.

**Comment by lemire** [comment]
@kf-vikas 

> I was hoping that the PR will contain a summary describing motivation/ existing problems for making this change. Since this change is too big, so it is beyond human review. The only way to review is to ensure that this commit hash matches from where it is picked up and it worked originally.

It is definitively too big for manual review, so we have to trust the code formatter (spotless).

The motivation for this change is as follows. We used to have a format checker which I removed because it was too limited. All it did was to check the format and issue a warning/failure when the format was inadequate. Furthermore, we limited it to only some subprojects.

With this PR, we will have both code format checks *and* automated consistent reformat over the whole code base.

As to why this is a nice feature: it is the only practical way to ensure that we have a common style throughout the code base. Having a fixed code formatting convention saves time in the long run because there is no more discussion on how to format the code. It is all automated.

**Review by kf-vikas — APPROVED** [review]
lgtm

**Review by anandheritage — APPROVED** [review]

**Comment by richardstartin** [comment]
I have no objections to reformatting the code

**Comment by fanson** [comment]
> Let us ping recent contributors:
> 
> @kf-vikas @fanson @anandheritage @iszhangpch @zoltanmeze
> 
> And also core contributors like @richardstartin
> 
> If anyone objects to a full code reformat, please speak up as this can be quite damaging if you have forked the project and are in the process of changing it...
> 
> There will be no going back once this PR is applied.

LGTM, I have no objections to reformatting the code too. 
We just need a standardized and common format that every commit follows.

**Comment by lemire** [comment]
I am merging this. Thanks to all the contributors and @blacelle in particular for making this work.

**Pull request merged**
